### PR TITLE
Tools: add junit output for autotest

### PR DESCRIPTION
**Pull request by khancyr**
Prototype to add junit xml report on autotest result. Need to install junitparser (pip install junitparser)

example with :  `Tools/autotest/autotest.py test.CopterTests2a --junit`

```
<?xml version='1.0' encoding='UTF-8'?>
<testsuites><testsuite name="ArduPilot Autotest" tests="2" errors="0" failures="0" skipped="0" time="370.806"><properties><property name="Autotest name" value="test.CopterTests2a"/></properties><testcase name="FixedYawCalibration" classname="Test Fixed Yaw Calibration" time="3.7409234046936035"/><testcase name="SITLCompassCalibration" classname="Test Fixed Yaw Calibration&quot;" time="367.0650143623352"/></testsuite></testsuites>
```

image

### CODE REVIEW

**Comment by Ryanf55** [comment]
Nice! Did you want to add that python package to the docker environments and also the environment setup scripts?

**Comment by khancyr** [comment]
Nope, I just did an initial implémentation but that should be easy to add as that is only one pip package deps.

Still some things to look at as the report name and locations. But it should be usable for testing

**Comment by Ryanf55** [comment]
Yep, perhaps we can have github parse it and display with one of these tools?
https://github.com/marketplace/actions/test-reporter

**Comment by khancyr** [comment]
cool ! The second one would be excellent I think ! That would give faster report and prevent to have to dig into the log for the failure ! @peterbarker interested ?

**Comment by nexton-winjeel** [comment]
Nice! We've got something similar. The other thing I'd suggest adding is a `timestamp` to the `testsuite` element.

Our version:
```python
    def output_xunit_results(self, test_results, skip_list):
        now = datetime.datetime.now().replace(microsecond=0)
        frame = self.vehicleinfo_key()
        suffix = ""
        # If we've only run one test (which happens when autotest is passed an explicit list of
        # subtests), put it in it's own file.
        if len(test_results) == 1 and len(skip_list) == 0:
            suffix = "-%s-%s" % (frame, test_results[0].test.name)

        xml_output = os.path.join(self.buildlogs_dirpath(), "autotest%s.xml" % suffix)
        self.progress("Writing test result in xUnit format to %s\n" % xml_output)

        with open(xml_output, "w") as f:
            f.write('<?xml version="1.0" encoding="UTF-8"?>\n')

            f.write('<testsuites name="Autotest" timestamp="%s">\n\n' % now.isoformat())
            f.write('    <testsuite name="%s Autotests">\n' % frame)

            for skipped in skip_list:
                (test, reason) = skipped
                f.write('        <testcase classname="%s" name="%s" time="0">\n' % (frame, test.name))
                f.write('            <skipped message="%s"/>\n' % reason)
                f.write('        </testcase>\n')

            for result in test_results:
                test = result.test
                time = self.test_timings.get(test.description, 0.0)
                if result.passed:
                    f.write('        <testcase classname="%s" name="%s" time="%f"/>\n' % (frame, test.name, time))
                else:
                    f.write('        <testcase classname="%s" name="%s" time="%f">\n' % (frame, test.name, time))
                    f.write('            <failure type="%s" message="see: %s"/>\n' % (result.exception, result.debug_filename))
                    f.write('        </testcase>\n')

            f.write('    </testsuite>\n\n')
            f.write('</testsuites>\n\n')
```

**Comment by Ryanf55** [comment]
One thing that would also be handy is to handle our known flapping tests. If the tests are known to be unreliable, is there anything other than skipping that we can do?

**Comment by khancyr** [comment]
For flapping tests there is already retry sequence

**Comment by khancyr** [comment]
For flapping tests there is already a retry sequence 

@nexton-winjeel nice thanks for the sharing, that is probably simpler than pulling a lib. Will look how to complete all this

File: Tools/autotest/common.py [hunk 1]
```diff
@@ -13488,9 +13501,41 @@ def autotest(self, tests=None, allow_skips=True):
...
         return len(self.fail_list) == 0
 
+    def create_junit_report(self, test_name: str, results: List[Result], skip_list: list[tuple[Test, dict[str, str]]]) -> None:
+        """Generate Junit report from the autotest results"""
+        from junitparser import TestCase, TestSuite, JUnitXml, Skipped, Failure
+        frame = self.vehicleinfo_key()
+        xml_filename = f"autotest_result_{frame}_{test_name}_junit.xml"
+        self.progress(f"Writing test result in jUnit format to {xml_filename}\n")
+
+        suites = TestSuite("ArduPilot Autotest")
```
Comment (nexton-winjeel) [inline Tools/autotest/common.py:13518]: Why the two test suites?

Comment (khancyr) [reply Tools/autotest/common.py:13518]: first one create the testsuites

Comment (nexton-winjeel) [reply Tools/autotest/common.py:13572]: This one is empty except for a timestamp?
```python
        suites = TestSuite("ArduPilot Autotest")
        suites.timestamp = datetime.now().replace(microsecond=0).isoformat()
```

This one gets the test cases added to it?
```python
        suite = TestSuite(f"Autotest {frame} {test_name}")
```

**Comment by Ryanf55** [comment]
> For flapping tests there is already a retry sequence
> 
> @nexton-winjeel nice thanks for the sharing, that is probably simpler than pulling a lib. Will look how to complete all this

Although yea it's possible to skip a library, it's at the cost of maintainability. I just refactored a bunch of rviz code to remove manual html building in a similar fashion. Libs are essentially free and improve readability.

File: Tools/autotest/common.py (hunk 1 above)
Comment (khancyr) [reply Tools/autotest/common.py:13572]: yes
the first one is the equivalent of your  : f.write('<testsuites name="Autotest" timestamp="%s">\n\n' % now.isoformat())

**Comment by khancyr** [comment]
Here I decide to keep the lib as I didn't want to reimplement myself the statistics update on timing and other stuffs. But we can probably skip it and use xml edit to update the fields as need...

For the github action reporter, it doesn't seems straight forward to implement. 
I did some try : https://github.com/khancyr/ardupilot/runs/17908021485 
But that need to split the  worfklow and make some weird results, so I will discard this part for now

File: Tools/autotest/common.py (hunk 1 above)
Comment (nexton-winjeel) [reply Tools/autotest/common.py:13572]: Except they're not equivalent:
* I'm putting the timestamp on the top-level `<testsuites>` element, e.g.
    ```xml
    <testsuites name="ArduPilot Autotest" timestamp="2023-10-23T10:12:07">
        <testsuite name="Autotest Frame Test">...</testsuite>
    </testsuites>
    ```
* You're putting the timestamp on a second `<testsuite>` element, e.g.
    ```xml
    <testsuites>
        <testsuite name="ArduPilot Autotest" timestamp="2023-10-23T10:12:07"/>
        <testsuite name="Autotest Frame Test">...</testsuite>
    </testsuites>
    ```

Comment (khancyr) [reply Tools/autotest/common.py:13572]: ha yes... doesn't work as intent ... from the lib code the first call create the testsuites, but it doesn't do it correctly ... annoying

**Comment by khancyr** [comment]
@Ryanf55 @nexton-winjeel hello, I have push a final version. I think will keep the lib for now. 

here is an output example : 
```
<?xml version="1.0" encoding="utf-8"?>
<testsuites>
	<testsuite name="Autotest ArduCopter test.CopterTests1a" timestamp="2023-11-16T14:43:23" tests="27" errors="0" failures="0" skipped="0" time="233.136">
		<testcase name="PIDTuning" classname="ArduCopter" time="5.101137638092041"/>
		<testcase name="ArmFeatures" classname="ArduCopter" time="20.81020736694336"/>
		<testcase name="SetHome" classname="ArduCopter" time="11.95091986656189"/>
		<testcase name="ConfigErrorLoop" classname="ArduCopter" time="2.7126312255859375"/>
		<testcase name="CPUFailsafe" classname="ArduCopter" time="5.354845762252808"/>
		<testcase name="Parameters" classname="ArduCopter" time="9.803567171096802"/>
		<testcase name="LoggerDocumentation" classname="ArduCopter" time="3.151670217514038"/>
		<testcase name="Logging" classname="ArduCopter" time="13.548582315444946"/>
		<testcase name="GetCapabilities" classname="ArduCopter" time="0.07680463790893555"/>
		<testcase name="InitialMode" classname="ArduCopter" time="2.385150909423828"/>
		<testcase name="NavDelayTakeoffAbsTime" classname="ArduCopter" time="7.005534887313843"/>
		<testcase name="NavDelayAbsTime" classname="ArduCopter" time="5.902580738067627"/>
		<testcase name="NavDelay" classname="ArduCopter" time="5.835424184799194"/>
		<testcase name="GuidedSubModeChange" classname="ArduCopter" time="7.143741607666016"/>
		<testcase name="MAV_CMD_CONDITION_YAW" classname="ArduCopter" time="14.092246770858765"/>
		<testcase name="LoiterToAlt" classname="ArduCopter" time="7.683399677276611"/>
		<testcase name="PayloadPlaceMission" classname="ArduCopter" time="9.601360082626343"/>
		<testcase name="PrecisionLoiterCompanion" classname="ArduCopter" time="7.640118598937988"/>
		<testcase name="Landing" classname="ArduCopter" time="6.856537818908691"/>
		<testcase name="PrecisionLanding" classname="ArduCopter" time="13.348423719406128"/>
		<testcase name="SetModesViaModeSwitch" classname="ArduCopter" time="0.9609272480010986"/>
		<testcase name="SetModesViaAuxSwitch" classname="ArduCopter" time="0.33498406410217285"/>
		<testcase name="AuxSwitchOptions" classname="ArduCopter" time="0.5513660907745361"/>
		<testcase name="AuxFunctionsInMission" classname="ArduCopter" time="3.2609589099884033"/>
		<testcase name="AutoTune" classname="ArduCopter" time="23.587228536605835"/>
		<testcase name="AutoTuneYawD" classname="ArduCopter" time="40.96126747131348"/>
		<testcase name="NoRCOnBootPreArmFailure" classname="ArduCopter" time="3.4743480682373047"/>
		<properties>
			<property name="Firmware Version Major" value="4"/>
			<property name="Firmware Version Minor" value="5"/>
			<property name="Firmware Version Rev" value="0"/>
			<property name="Firmware hash" value="d3d0f91c"/>
			<property name="Git hash" value="d3d0f91c92a8d76a42d6fd7aa5ecb398bb464d6a"/>
			<property name="Mavproxy Version Major" value="1"/>
			<property name="Mavproxy Version Minor" value="8"/>
			<property name="Mavproxy Version Rev" value="67"/>
		</properties>
	</testsuite>
</testsuites>

```

**Comment by khancyr** [comment]
[autotest_result_ArduCopter_test.CopterTests1a_junit.xml.pdf](https://github.com/ArduPilot/ardupilot/files/13379187/autotest_result_ArduCopter_test.CopterTests1a_junit.xml.pdf)

File: .dockerignore [hunk 2]
```diff
@@ -28,3 +28,4 @@ mav.*
 !Tools/environment_install/install-prereqs-ubuntu.sh
 !Tools/environment_install/install-prereqs-arch.sh
 !Tools/completion
+autotest_result_*_junit.xml
```
Comment (Ryanf55) [inline .dockerignore:31]: Please run pre-commit to catch this missing newline

File: Tools/autotest/vehicle_test_suite.py [hunk 3]
```diff
@@ -1540,6 +1547,14 @@ def __init__(self,
...
         self.build_opts = build_opts
         self.num_aux_imus = num_aux_imus
+        self.generate_junit = generate_junit
+        if generate_junit:
+            try:
+                spec = importlib.util.find_spec("junitparser")
+                if spec is None:
+                    raise ImportError
+            except ImportError as e:
+                raise ImportError(f"Junit export need junitparser package.\n {e} \nTry: python -m pip install junitparser")
```
Comment (Ryanf55) [inline Tools/autotest/vehicle_test_suite.py:1557]:
```suggestion
                raise ImportError(f"Junit export need junitparser package.\n {e} \nTry: python3 -m pip install junitparser")
```

`python` is no longer the right program name on 22.04, you need to use `python3`

**Review by Ryanf55 — CHANGES_REQUESTED** [review]
Just a few small things, otherwise looks great! The PDF has a good amount of info.

**Comment by khancyr** [comment]
pdf was generated with junit2html and then converted from html to pdf as github don't allow to post it.

I will fix CI and it should be good

File: Tools/autotest/pysim/util.py [hunk 4]
```diff
@@ -815,6 +815,13 @@ def load_local_module(fname):
     return ret
 
 
+def get_git_hash():
+    githash = run_cmd('git rev-parse HEAD', output=True, directory=reltopdir('.')).strip()
```
Comment (nexton-winjeel) [inline Tools/autotest/pysim/util.py:819]:
```suggestion
    githash = run_cmd('git rev-parse --short=8 HEAD', output=True, directory=reltopdir('.')).strip()
```

Maybe use `--short=8` so the repo hash is the same length as the firmware hash?

**Comment by nexton-winjeel** [comment]
Looks good to me!

File: Tools/autotest/pysim/util.py (hunk 4 above)
Comment (khancyr) [reply Tools/autotest/pysim/util.py:819]: yep, will make comparison easier, thx

**Review by Ryanf55 — APPROVED** [review]

**Review by andyp1per — APPROVED** [review]

**Pull request merged**